crates/bevy_asset/src/temp.rs

```diff
@@ -0,0 +1,318 @@
+    directory: TempDirectoryKind,
+}
+
+impl TempDirectory {
+    /// Try to create a new [`TempDirectory`] resource, which uses a randomly created
+    /// directory in the user's temporary directory. This can fail if the platform does not
+    /// provide an appropriate temporary directory, or the directory itself could not be created.
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn new_transient() -> std::io::Result<Self> {
+        let directory = TempDirectoryKind::new_transient()?;
+
+        Ok(Self { directory })
+    }
+
+    /// Create a new [`TempDirectory`] resource, which uses a provided directory to store temporary
+    /// assets. It is assumed this directory already exists, and it will _not_ be deleted on exit.
+    pub fn new_persistent(path: impl Into<PathBuf>) -> Self {
+        let directory = TempDirectoryKind::new_persistent(path);
+
+        Self { directory }
+    }
+
+    /// Get the [`Path`] to the directory used for temporary assets.
+    pub fn path(&self) -> &Path {
+        self.directory.path()
+    }
+
+    /// Persist the current temporary asset directory after application exit.
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn persist(&mut self) -> &mut Self {
+        self.directory.persist();
+
+        self
+    }
+}
+
+/// Private resource to store the temporary directory used by `temp://`.
+/// Kept private as it should only be removed on application exit.
+enum TempDirectoryKind {
+    /// Uses [`TempDir`](tempfile::TempDir)'s drop behavior to delete the directory.
+    /// Note that this is not _guaranteed_ to succeed, so it is possible to leak files from this
+    /// option until the underlying OS cleans temporary directories. For secure files, consider using
+    /// [`tempfile`](tempfile::tempfile) directly.
+    #[cfg(not(target_arch = "wasm32"))]
+    Delete(tempfile::TempDir),
+    /// Will not delete the temporary directory on exit, leaving cleanup the responsibility of
+    /// the user or their system.
+    Persist(PathBuf),
+}
+
+impl TempDirectoryKind {
+    #[cfg(not(target_arch = "wasm32"))]
+    fn new_transient() -> std::io::Result<Self> {
+        let directory = tempfile::TempDir::with_prefix("bevy_")?;
+        Ok(Self::Delete(directory))
+    }
+
+    fn new_persistent(path: impl Into<PathBuf>) -> Self {
+        Self::Persist(path.into())
+    }
+
+    fn path(&self) -> &Path {
+        match self {
+            #[cfg(not(target_arch = "wasm32"))]
+            Self::Delete(x) => x.as_ref(),
+            Self::Persist(x) => x.as_ref(),
+        }
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    fn persist(&mut self) -> &mut Self {
+        let mut swap = Self::Persist(PathBuf::new());
+
+        std::mem::swap(self, &mut swap);
+
+        let new = match swap {
+            Self::Delete(x) => Self::Persist(x.into_path()),
+            x @ Self::Persist(_) => x,
+        };
+
+        *self = new;
+
+        self
+    }
+}
+
+pub(crate) fn get_temp_source(
```

Comment:
I do agree there's not a lot of utility to the `TempDirectory` type on web. I think I'll leave it in place tho just as it's part of the user-facing configuration of the `temp://` source, and I want to leave the OPFS independent of the `temp://` feature, since users may find it useful to manage assets in the browser as well.